File: dart/lib/src/bindings/burn_domain.dart
import 'package:solana/solana.dart' hide RpcClient;

import '../constants/addresses.dart';
import '../instructions/instructions.dart';
import '../utils/derive_address.dart';
import '../utils/get_reverse_address_from_domain_address.dart';
import '../utils/base58_utils.dart';

/// Parameters for burning a domain
class BurnDomainParams {
  const BurnDomainParams({
    required this.domain,
    required this.owner,
    required this.refundAddress,
  });

  /// The domain name to burn
  final String domain;

  /// The current owner's address
  final String owner;

  /// The address to refund rent to
  final String refundAddress;
}

/// Creates a domain burn instruction
///
/// Burns a domain and refunds rent to the specified address.
/// This matches the js-kit/src/bindings/burnDomain.ts implementation.
Future<TransactionInstruction> burnDomain(BurnDomainParams params) async {
  // Derive the domain address
  final domainAddress = await deriveAddress(params.domain);

  // Derive proper addresses using correct PDA derivation patterns
  final reverseAddress =
      await getReverseAddressFromDomainAddress(domainAddress);

  // State address is derived as PDA with domain address as seed
  final stateAddress = await Ed25519HDPublicKey.findProgramAddress(
    seeds: [Base58Utils.decode(domainAddress)],
    programId: Ed25519HDPublicKey.fromBase58(registryProgramAddress),
  ).then((result) => result.toBase58());

  // For reselling state, use similar pattern - this would need to match JS implementation
  final resellingStateAddress =
      stateAddress; // Often same as state for burn operations

  // Create burn instruction parameters
  final burnParams = BurnDomainInstructionParams(
    nameServiceId: nameProgramAddress,
    systemProgram: systemProgramAddress,
    domainAddress: domainAddress,
    reverse: reverseAddress,
    resellingState: resellingStateAddress,
    state: stateAddress,
    centralState: centralState,
    owner: params.owner,
    target: params.refundAddress,
    programAddress: registryProgramAddress,
  );

  // Create and build the burn instruction
  final burnInstr = BurnDomainInstruction(
    params: burnParams,
  );

  return burnInstr.build();
}

File: js/src/bindings/burnDomain.ts
import { PublicKey, SystemProgram } from "@solana/web3.js";

import {
  NAME_PROGRAM_ID,
  REGISTER_PROGRAM_ID,
  REVERSE_LOOKUP_CLASS,
} from "../constants";
import { BurnInstruction } from "../instructions/burnInstruction";
import { getSnsDomainKeySync } from "../utils/getSnsDomainKeySync";
import { getReverseKeySync } from "../utils/getReverseKeySync";
import { _parseSnsTopLevelDomain } from "../utils/parseSnsDomain";

/**
 * Builds an instruction to burn a top-level `.sns` domain and its reverse lookup account.
 *
 * @param domain Full `.sns` domain name
 * @param owner Current owner of the domain
 * @param target Account that receives reclaimed lamports
 * @returns Transaction instruction.
 */
export const burnDomain = (
  domain: string,
  owner: PublicKey,
  target: PublicKey,
) => {
  const trimmedDomain = _parseSnsTopLevelDomain(domain);

  const { pubkey } = getSnsDomainKeySync(trimmedDomain);
  const [state] = PublicKey.findProgramAddressSync(
    [pubkey.toBuffer()],
    REGISTER_PROGRAM_ID,
  );
  const [resellingState] = PublicKey.findProgramAddressSync(
    [pubkey.toBuffer(), Uint8Array.from([1, 1])],
    REGISTER_PROGRAM_ID,
  );

  const ix = new BurnInstruction().getInstruction(
    REGISTER_PROGRAM_ID,
    NAME_PROGRAM_ID,
    SystemProgram.programId,
    pubkey,
    getReverseKeySync(trimmedDomain),
    resellingState,
    state,
    REVERSE_LOOKUP_CLASS,
    owner,
    target,
  );
  return ix;
};
